src/cargo/core/compiler/context/unit_dependencies.rs

```diff
@@ -29,8 +29,13 @@ pub fn build_unit_dependencies<'a, 'cfg>(
 ) -> CargoResult<HashMap<Unit<'a>, Vec<Unit<'a>>>> {
     let mut deps = HashMap::new();
     for unit in roots.iter() {
-        // Dependencies of tests should not have `panic` set.
-        let profile_for = if unit.mode.is_any_test() {
+        // Dependencies of tests/benches should not have `panic` set.
+        // We check the global test mode to see if we are running in `cargo
+        // test` in which case we ensure all dependencies have `panic`
+        // cleared, and avoid building the lib thrice (once with `panic`, once
+        // without, once for --test).  In particular, the lib included for
+        // doctests and examples are `Build` mode here.
+        let profile_for = if unit.mode.is_any_test() || cx.build_config.test {
```

Comment:
Hm, I am genuinely not sure what the correct behavior would be here? Suppose we are compiling the integration test, which only executes the binary in the same package. It is reasonable to assume that the binary might be compiled with `panic=abort`, but the test itself should be compiled with `panic=unwind`, so common dependencies should be compiled twice. 

Perhaps the best path forward is to preserve the current, buggy, behavior, land this PR, and then fix the behavior in a follow-up PR?